### ISSUE TITLE
Environment Bikeshed

### ISSUE THREAD

**Issue by ThePhD** [ThePhD/sol2#387]
Before it was asked if Environments could be done by sol2. The answer before was No, because it was difficult to do.

sol2's latest [now has environments](https://github.com/ThePhD/sol2/blob/develop/examples/environments.cpp).

Is that all that there needs to be? Are the [docs](http://sol2.readthedocs.io/en/latest/api/environment.html) good enough? It seems to work with both Lua 5.1/5.2/5.3 and LuaJIT as the tests show, but is there any power we're missing?

This issue won't stay open for long, so anyone who wants to chime in should do it fast (I remember @squeek502 asking about this).

**Comment by squeek502** [comment]
Nice, looks good. A few things:

---

Returns from functions with a different environment seem to not work correctly (only tested with Lua 5.1). Test code:

```c++
TEST_CASE("environment return values")
{
	sol::state lua;
	lua.open_libraries(sol::lib::base);

	lua.script("a = function() return 5 end");

	sol::function a = lua["a"];

	int plainReturn = a();
	REQUIRE(plainReturn == 5);

	sol::environment env(lua, sol::create);
	sol::set_environment(a, env);

	int envReturn = a();
	REQUIRE(envReturn == 5);
}
```
Result:
```
envs.cpp(65): FAILED:
  REQUIRE( envReturn == 5 )
with expansion:
  0 == 5
```

---

It would be nice to add an example of `set_environment` used outside of the `script/do_x` functions to the examples/tests/docs.

Something like (given the above issue is fixed):

```c++
TEST_CASE("set_environment")
{
	sol::state lua;
	lua.open_libraries(sol::lib::base);

	SECTION("retrieving a value set in the environment before execution")
	{
		lua.script("a = function() return test end");

		sol::function a = lua["a"];
		sol::environment env(lua, sol::create);
		env["test"] = 5;
		sol::set_environment(a, env);

		// the function returns the value from the environment table
		int result = a();
		REQUIRE(result == 5);
	}

	SECTION("retrieving a value set in the environment during execution")
	{
		lua.script("a = function() test = 5 end");

		sol::function a = lua["a"];
		sol::environment env(lua, sol::create);
		sol::set_environment(a, env);

		a();

		// the value can be retrieved from the env table
		int result = env["test"];
		REQUIRE(result == 5);

		// the global environment is not polluted
		auto globalValue = lua["test"];
		REQUIRE(!globalValue.valid());
	}
}
```

**Comment by squeek502** [comment]
One more thing: it's unclear to me and in the docs what the `sol::new_table nt` argument of the `environment` constructor can be used for. All example code and the docs simply pass `sol::create`, so it's not clear what the alternative is. The [table constructor docs](http://sol2.readthedocs.io/en/latest/api/table.html?highlight=new_table#members) are similarly vague, and `new_table` has [no usages or examples in the codebase](https://github.com/ThePhD/sol2/search?utf8=%E2%9C%93&q=new_table&type=).

I guess what I'm getting at is it seems like either: 1) there should be examples of how `new_table` can be used (therefore justifying it as a required argument), or 2) the `new_table` argument should be defaulted to `sol::create` and put after `fallback`, so that it's usage is more in line with its seemingly rare use case.

Also worth noting that the following code compiles but segfaults:

```
sol::state lua;
lua.script("a = function() test = 5 end");

sol::function a = lua["a"];
sol::environment env(lua);
sol::set_environment(a, env);
```

**Comment by ThePhD** [comment]
`new_table` has no usages or examples because it's only purpose is to signal to the framework that you want to create a new table, generally VIA `sol::create`. In the search you performed, the first result is in `environment.hpp`, where we create a new table using `sol::stack_table(L, sol::new_table(0, 1), ...` Overloads are also ambiguous if we don't have a marker in the 2nd positions, plus more than a table can be used as an environment (In Lua 5.2+, anyways, where the rules are relaxed and someone can use a userdata as an enviroment fallback lookup or as the environment itself).

I can still add some description to `new_table` to make it more apparent why it's necessary.

The code that compiles and segfaults is allowed. Someone can create an invalid environment that has no actual reference stored in Lua (registry_index == LUA_REFNIL), but still has the state set on it.

**Comment by ThePhD** [comment]
The code for setting an environment doesn't work when the target is a function for some reason, and only in Lua 5.1. The manual gives no reason why this would be the case: lua_setfenv should be exactly what's needed there. There's also a vague reference in the 5.1 manual that the environment for a (C) function is stored at LUA_ENVIRONINDEX. Chances are maybe that I need to manipulate that pseudo-index in the upvalues to change the function's upvalue later on...?

If I can get it to work, I'll be happy to put that code in the samples.

**Comment by squeek502** [comment]
> The code for setting an environment doesn't work when the target is a function for some reason, and only in Lua 5.1.

This seems to be specific to sol2. The following code works:

```c++
TEST_CASE("environment return values using lua_setfenv")
{
	sol::state lua;
	lua.open_libraries(sol::lib::base);

	lua.script("a = function() return 5 end");

	sol::function a = lua["a"];

	int plainReturn = a();
	REQUIRE(plainReturn == 5);

	sol::table env = lua.create_table();

	a.push();
	env.push();
	lua_setfenv(lua, -2);

	int envReturn = a();
	REQUIRE(envReturn == 5);
}
```

**Comment by ThePhD** [comment]
@squeek502 Sorry about fixing it at the same time, but I did push the fix and now your code works in Lua 5.1. Plus, I also upgraded / updated Travis, so hopefully things shouldn't just explode and break. Also, I swapped the order of `set_environment`, and added a member function `env.set_on( thingy );`.

I also fixed a bug @HoraceWeebler reported on the Gitter chat.

Finally, beefed up the docs where they were weak: 
sol2.readthedocs.io/en/latest/api/environment.html
http://sol2.readthedocs.io/en/latest/api/new_table.html

Let me know what you think.
